lib/IRGen/GenStruct.cpp

```diff
@@ -367,6 +390,55 @@ namespace {
       (void)ClangDecl;
     }
 
+    void destroy(IRGenFunction &IGF, Address address, SILType T,
+                 bool isOutlined) const override {
+      auto *destructor = getCXXDestructor(T);
+      // If the destructor is trivial, clang will assert when we call
+      // `emitCXXDestructorCall` so, just let Swift handle this destructor.
+      if (!destructor || destructor->isTrivial()) {
+        // If we didn't find a destructor to call, bail out to the parent
+        // implementation.
+        StructTypeInfoBase<AddressOnlyClangRecordTypeInfo, FixedTypeInfo,
+                           ClangFieldInfo>::destroy(IGF, address, T,
+                                                    isOutlined);
+        return;
+      }
+
+      if (!destructor->isUserProvided()) {
+        assert(!destructor->isDeleted() &&
+               "Swift cannot handle a type with no known destructor.");
+        // Make sure we define the destructor so we have something to call.
+        auto &sema = IGF.IGM.Context.getClangModuleLoader()->getClangSema();
+        sema.DefineImplicitDestructor(clang::SourceLocation(), destructor);
+      }
+
+      clang::GlobalDecl destructorGlobalDecl(destructor, clang::Dtor_Complete);
+      auto *destructorFnAddr = cast<llvm::Function>(
+          IGF.IGM
+              .getAddrOfClangGlobalDecl(destructorGlobalDecl, NotForDefinition)
+              ->stripPointerCasts());
```

Comment:
So, I'll explain my original thinking, then I'll explain why it's wrong 😁 

It's really important that this is an `llvm::Function` so that we can get it's arg types, etc. below. But, there are no guarantees that we get an `llvm::Function` back from `getAddrOfClangGlobalDecl` (or `clang::CodeGenerator::GetAddrOfGlobal` which is the function it's implemented with) so, I was worried we might get a bitcast instruction or a GEP instruction of the first element. If we got either of those (or something else) it wouldn't be an `llvm::Function` (it would be an `llvm::Value`, maybe pointing to an `llvm::Function`), so there'd be an issue. 

But, now I don't think we actually need it (for two reasons). First, it's guaranteed that we'll get back an `llvm::Constant` and the only `llvm::Constant` that I can think of that eventually points to an `llvm::Function` would be an `llvm::Function` itself. Second, and a worse justification because this code may change in the future, following `GetAddrOfGlobal`s path for structors, the only possible return value is an `llvm::Function`.

Anyway, I'll remove the call. But does that make sense?

Reply:
Yes it does, thank you!